Multisample framebuffer

mgsx-dev opened the pull request:
This is an implementation proposal for framebuffer multisample feature.

When using multisampling with framebuffers, you need 2 OpenGL FBOs : 
* a mutisample FBO used to capture the scene. This one can't be rendered directly or used as regular textures.
* a non multisample FBO where resolved samples will be copied to textures and then usable as regular textures.

There was several ways to implementing it. There was an old PR where GLFrameBuffer was handling 2 OpenGL framebuffers. This overcomplexify GLFrameBuffer code with a lot of code duplication. It also makes it more complex to configure and introduce confusion : (eg. which one is bound when calling bind method). So i decided to not take that route.

This PR use an approach closer to the OpenGL API. As you can see in the test, user code have to create 2 FrameBuffers and call the transfer method to copy multisamples into the other FBO. Another benefit is that in some case (hardware specific) it's not possible to have a texture for the stencil buffer or the packed depth and stencil buffer. With this approach, you can render a scene using both depth and stencil buffers, and later transfer only depth pixels to a texture. It also allows internal format conversion (eg. rendering colors into a multisample float buffer and transfer to an RGBA8 texture).

Note that, it could be possible to have texture attachments with multisample enabled. That allows to get hand on samples in order to compose with other multisampled textures and handle antialiasing manually. These textures can't be rendered directly neither. This feature (a very advanced case imho) would requires more changes and i thought it didn't worth it at the moment. An exception is thrown when using multisample FBO and texture attachement together.

Transfer code is a bit complex because of OpenGL API apparently designed for non MRT context. `glBlitFramebuffer` is the command that perform the copy, it allows to copy depth, stencil and one color buffer in a single call. But if you have multiple color attachement, it's necessary to call it again for other color buffers.

There are surely some stuff to improve, i'd like to have your opinion on the API proposal.

mgsx-dev commented:
> When running the `GL31FrameBufferMultisampleMRTTest` on Windows Desktop I am just seeing a black screen. Is this normal?

No, it should looks like this : 
![Screenshot 2023-07-21 04:41:21](https://github.com/libgdx/libgdx/assets/8074238/424b7eee-8495-4146-95aa-d19fc2b85559)

Maybe on Windows it requires `glEnable(GL_MULTISAMPLE);` ? from [this tutorial](https://learnopengl.com/Advanced-OpenGL/Anti-Aliasing) : 

> On most OpenGL drivers, multisampling is enabled by default so this call is then a bit redundant, but it's usually a good idea to enable it anyways.

Could you try to do that ? it should be done in desktop code, because it's not part of GLES : `glEnable(GL13.GL_MULTISAMPLE);`

JamesTKhan commented:
> 

No luck, I did try that yesterday too. Tried again just to be sure, tried both GL13 ENUM as well as just plugging its integer value in directly.
`Gdx.gl.glEnable(GL13.GL_MULTISAMPLE);`
`Gdx.gl.glEnable(0x809D);` 

I also tried putting GLProfiler into the test, but not getting any error messages out.

On Windows 11 with Nvidia 3080

![image](https://github.com/libgdx/libgdx/assets/28971753/5a99109c-cd3d-4565-bae4-e50a712c1f8a)

JamesTKhan commented:
Narrowed the problem down. If I remove `Gdx.gl.glEnable(GL20.GL_DEPTH_TEST);` in render method of test, it works.

mgsx-dev commented:
> Narrowed the problem down. If I remove `Gdx.gl.glEnable(GL20.GL_DEPTH_TEST);` in render method of test, it works. Or I can keep the depth test, but I must add true to the first ScreenUtils clear method to clear depth as the first one does not pass true.
> 
> `ScreenUtils.clear(Color.CLEAR, true);`

Good catch! it's fixed. I remember now that default frame buffer clearing have difference between windows and linux. Thanks.

JamesTKhan commented:
> > Narrowed the problem down. If I remove `Gdx.gl.glEnable(GL20.GL_DEPTH_TEST);` in render method of test, it works. Or I can keep the depth test, but I must add true to the first ScreenUtils clear method to clear depth as the first one does not pass true.
> > `ScreenUtils.clear(Color.CLEAR, true);`
> 
> Good catch! it's fixed. I remember now that default frame buffer clearing have difference between windows and linux. Thanks.

Just checking back in. Your latest changes are working for me on the test. I submitted my other two concerns/observations as part of the GH review process.

xzaleksey commented:
Any updates?)

antzGames commented:
Hey folks.  I check in every week to see if they are any updates on this PR.  I am sure many are also waiting on this.  

I know its summer and we are all busy.   A friendly reminder that the community appreciates the work you do.  Cheers!

antzGames commented:
What can the community do to help the libGDX contributors reduce the cycle time for PRs to get merged?  This is an honest and sincere question.

tommyettinger commented:
> What can the community do to help the libGDX contributors reduce the cycle time for PRs to get merged? This is an honest and sincere question.

I don't know enough about this domain to be able to adequately review this PR. I think we're kind-of short on people who do have that domain knowledge, much like we are with people who understand Tiled PRs.

JamesTKhan commented on gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java:
```diff
@@ -350,6 +401,71 @@ public void end (int x, int y, int width, int height) {
+		}
+
+		// restore draw buffers for destination
+		Gdx.gl30.glDrawBuffers(destination.defaultDrawBuffers.limit(), destination.defaultDrawBuffers);
```
When testing out the FBO transfer method, I was getting a NPE crash if transferring to a regular FBO that is not MRT.  `defaultDrawBuffers` is only available on FBO that has MRT or samples enabled. 

Locally I added a null check like so, and the transfer now works for non-MRT destination FBOS

```
if (destination.defaultDrawBuffers != null) {
	// restore draw buffers for destination
	Gdx.gl30.glDrawBuffers(destination.defaultDrawBuffers.limit(), destination.defaultDrawBuffers);
}
```

mgsx-dev replied:
indeed, i updated GL31FrameBufferMultisampleTest to use the real implementation and confirmed the issue. It's fixed now.

obigu commented:
@mgsx-de It'd be great if you could have a look at @JamesTKhan review comment https://github.com/libgdx/libgdx/pull/7185#discussion_r1345826056 when you have the chance, considering the interest we should aim at getting it merged in 1.12.2.

mgsx-dev commented on gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java:
```diff
@@ -350,6 +401,71 @@ public void end (int x, int y, int width, int height) {
+				singleInt.flip();
+				Gdx.gl30.glDrawBuffers(1, singleInt);
+
+				Gdx.gl30.glBlitFramebuffer(0, 0, getWidth(), getHeight(), 0, 0, destination.getWidth(), destination.getHeight(),
```
indeed, i added a size check and throw an error

mgsx-dev commented on gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java:
```diff
@@ -350,6 +401,71 @@ public void end (int x, int y, int width, int height) {
+			if (attachment.isDepth && bufferBuilder.hasDepthRenderBuffer || bufferBuilder.hasPackedStencilDepthRenderBuffer) {
+				copyBits |= GL20.GL_DEPTH_BUFFER_BIT;
+			} else if (attachment.isStencil && bufferBuilder.hasStencilRenderBuffer
+				|| bufferBuilder.hasPackedStencilDepthRenderBuffer) {
```
indeed, the conditions were wrong, it's fixed now. Nice catch!

mgsx-dev commented:
@JamesTKhan i think i addressed all your points (thank you for the review), it seems that you need to review again unblock the merge.

@obigu I think it's ready to be merged.

Thank you all for your patience

JamesTKhan reviewed (dismissed):
Thank you!
